packages/zoe/zoe.js

```diff
@@ -42,137 +35,238 @@ const makeZoe = async (additionalEndowments = {}) => {
+
+        const payoutRuleMatrix = offers.map(offer => offer.payoutRules);
+        const currentExtentMatrix = offers.map(offer => offer.extents);
+        const extentOpsArray = assayTable.getExtentOpsForAssays(assays);
 
         // 1) ensure that rights are conserved
         insist(
-          areRightsConserved(extentOpsArray, currentExtents, reallocation),
+          areRightsConserved(
+            extentOpsArray,
+            currentExtentMatrix,
+            newExtentMatrix,
+          ),
         )`Rights are not conserved in the proposed reallocation`;
 
         // 2) ensure 'offer safety' for each player
         insist(
-          isOfferSafeForAll(extentOpsArray, payoutRulesArray, reallocation),
+          isOfferSafeForAll(extentOpsArray, payoutRuleMatrix, newExtentMatrix),
         )`The proposed reallocation was not offer safe`;
 
         // 3) save the reallocation
-        adminState.setExtentsFor(offerHandles, reallocation);
+        //    Note: We update both extents and units so that units can be
+        //    used in the future.
+        offerTable.updateExtentMatrix(offerHandles, newExtentMatrix);
+        const unitOpsArray = assayTable.getUnitOpsForAssays(assays);
+        const newUnitMatrix = newExtentMatrix.map(extentsRow =>
+          extentsRow.map((extent, i) => unitOpsArray[i].make(extent)),
+        );
+        offerTable.updateUnitMatrix(offerHandles, newUnitMatrix);
       },
 
       /**
-       * The contract can "complete" an offer to remove it
-       * from the ongoing contract and resolve the
-       * `result` promise with the player's payouts (either winnings or
-       * refunds). Because Zoe only allows for reallocations that
-       * conserve rights and are 'offer-safe', we don't need to do
-       * those checks at this step and can assume that the
-       * invariants hold.
+       * The contract can "complete" an offer to remove it from the
+       * ongoing contract and resolve the player's payouts (either
+       * winnings or refunds). Because Zoe only allows for
+       * reallocations that conserve rights and are 'offer-safe', we
+       * don't need to do those checks at this step and can assume
+       * that the invariants hold.
        * @param  {object[]} offerHandles - an array of offerHandles
        */
-      complete: offerHandles =>
-        completeOffers(adminState, readOnlyState, offerHandles),
+      complete: completeOffers,
 
       /**
-       *  The contract can create an empty offer and get the
-       *  associated offerHandle. This allows the contract to use this
-       *  offer slot for record-keeping. For instance, to represent a
-       *  pool, the contract can create an empty offer and then
-       *  reallocate other extents to this offer.
+       * The contract can create an "empty" offer. This is used by the
+       * autoswap contract to create an offer representing the
+       * liquidity pool balances.
        */
       escrowEmptyOffer: () => {
```

Comment:
Hard to read this an be certain that it is equivalent to a degenerate use of `escrowOffer`. Would prefer that it be coded as a wrapper around a call to `escrowOffer` so that would be clear.

Separately, we talked about getting rid of all methods on the facet that are redundant with methods on the zoeService, of which the escrowing methods stand out. Instead, we could add one `getZoeService()` method to the facet, so that a contract can generally act as a client of the service in any way it wants. However, we'll do that after this PR. (Worth adding a TODO or issue or something.)

When that happens, `escrowEmptyOffer()` would move to the service, but still as a trivial wrapper. Or perhaps a separate convenience function, keeping unnecessary code out of Zoe itself.

Reply:
I factored out a method called `depositPayments` so that the duplicate code is de-duped. I like the `getZoeService` idea a lot.